IDE section (#372)
cyberbotics/webots: docs/guide/using-your-ide.md

## Using Your IDE

> Using an [Integrated Development Environments (IDE)](https://en.wikipedia.org/wiki/Integrated_development_environment) is for sure convenient for its features (graphical debugger, edition tools, etc.).

Reviewer: ```suggestion
Using an [Integrated Development Environment (IDE)](https://en.wikipedia.org/wiki/Integrated_development_environment) is convenient as it generally features advanced source code edition tools, a built-in debugger, etc.
```

> You may be interested in using your favorite IDE to develop a Webots controller.

Reviewer: ```suggestion
Fortunately, you can use your favorite IDE to develop a Webots controller.
```

> A priori, Webots can support any IDE to create, build and debug its controllers.

Reviewer: ```suggestion
A priori, Webots works with any IDE to create, build and debug robot controllers.
```

It is simply a matter of setting up the IDE correctly to fulfill the Webots controller build rules.

Documenting every IDE for each OS is a huge and unmaintainable task.
This is why only some of them are documented.
They could serve as a reference for others.

The build rules are:

- The target executable file should have the same name as the controller directory, and follow strictly this path: `$WEBOTS_PROJECT/controllers/$(CONTROLLER_NAME)/$(CONTROLLER_NAME)$(EXE_EXTENSION)`.

> - The executable should be linked with the Webots controller library (and eventually include the corresponding path):

Reviewer: Unfortunately, this is not sufficient with Visual C++ where we need to recompile all the libCppController files with the controller...

%tab-component "language"

%tab "C"

```
libraries:
- $(WEBOTS_HOME)/lib/$(SL_PREFIX)Controller$(SL_SUFFIX)

includes:
- $(WEBOTS_HOME)/include/controller/c
```

%tab-end

%tab "C++"

```
libraries:
- $(WEBOTS_HOME)/lib/$(SL_PREFIX)Controller$(SL_SUFFIX)
- $(WEBOTS_HOME)/lib/$(SL_PREFIX)CppController$(SL_SUFFIX)

includes:
- $(WEBOTS_HOME)/include/controller/cpp
```

%tab-end

%tab "Java"

```
libraries:
- $(WEBOTS_HOME)/lib/java/Controller.jar
```

%tab-end

%tab "Python"

```
N/A because Python is interpreted.
```

%tab-end

%end

---

*Legend*

- `WEBOTS_HOME` is the path to Webots.
- `WEBOTS_PROJECT` is the path to your Webots project.
- `CONTROLLER_NAME` is the name of your controller.
- `EXE_EXTENSION` is the executable file suffix.

>     - For C or C++: `.exe` on Windows, and nothing else.

Reviewer: ```suggestion
    - For C or C++: `.exe` on Windows, and an empty string on other operating systems.
```

    - For Python: `.py`.
    - For Java: `.class` or `.jar`.

> - `SL_PREFIX` is the prefix of a shared library: `lib` on Linux or macOS, and nothing on Windows.

Reviewer: ```suggestion
- `SL_PREFIX` is the prefix of a shared library: `lib` on Linux or macOS, and an empty string on Windows.
```

- `SL_SUFFIX` is the suffix of a shared library: `.so` on Linux, `.dylib` on macOS and `.dll` on Windows.
- `PYTHON_VERSION` is your Python version, but concatenated (`27`, `37`, etc.).

---

### Visual Studio

On Windows, [Visual Studio](https://visualstudio.microsoft.com) can be used to create a C or C++ controller.

A `Visual Studio` project can be simply created using the "New Robot Controller..." wizard.
To do so, you should simply select the **Wizards / New Robot Controller...** menu item, choose the C or C++ language, and select a Visual Studio project.

> The target project can be open in `Visual Studio`.

Reviewer: It's a pity to drop useful bits of information about debugging a controller with Visual Studio or adding the car and driver libraries...

Author: Would it be ok it I try to re-introduce:
- 8th item from https://cyberbotics.com/doc/guide/using-visual-studio-with-webots#configuration
- vehicle/driver parts from https://cyberbotics.com/doc/guide/using-visual-studio-with-webots#link-with-the-webots-libraries

?

Reviewer: Yes, that would be certainly helpful.

### CMake

[CMake](https://cmake.org) is a cross-platform free and open-source software tool for managing the build process of software using a compiler-independent method.
Using [its generators](https://cmake.org/cmake/help/v3.0/manual/cmake-generators.7.html), it generates native build environments, such as `XCode`, `CodeBlocks`, `Sublime Text 2` or `Eclipse` projects.
The actual build is processed in these environments.

> For the `Visual Studio` target, prefer the [solution above](#visual-studio).

Reviewer: ```suggestion
For the `Visual Studio` target, prefer the [above solution](#visual-studio).
```

As a template, you could copy the following `CMakeLists.txt` file to your controller directory.
This template is only a sample, it may be adapted depending on your CMake target.

> ```cmake

Reviewer: Instead of this long piece of configuration file, wouldn't it be simpler to add a wizard for cmake? That would also be more consistent with other IDEs, such as Visual Studio.

ELSE()
  link_directories($ENV{WEBOTS_HOME}/lib)
ENDIF()
set (LIBRARIES ${CMAKE_SHARED_LIBRARY_PREFIX}Controller${CMAKE_SHARED_LIBRARY_SUFFIX} ${CMAKE_SHARED_LIBRARY_PREFIX}CppController${CMAKE_SHARED_LIBRARY_SUFFIX})
include_directories($ENV{WEBOTS_HOME}/include/controller/c $ENV{WEBOTS_HOME}/include/controller/cpp)

# Setup the target executable.
add_executable(${PROJECT} ${SOURCES})
target_link_libraries(${PROJECT} ${LIBRARIES})

# Copy the target executable at the right location.
add_custom_command(TARGET ${PROJECT} POST_BUILD COMMAND ${CMAKE_COMMAND} -E
  copy ${CMAKE_BINARY_DIR}/${PROJECT} ${CMAKE_SOURCE_DIR}
)
```

Then, create a `build` directory, create the native project, and build it in the target IDE.
For example:

```shell
# export WEBOTS_HOME=...
cd $WEBOTS_PROJECT/controllers/$CONTROLLER_NAME
# edit CMakeLists.txt
mkdir build
cd build
cmake .. -G "Unix Makefiles"
make
```

### Qt Creator

[Qt Creator](https://www.qt.io) is a cross-platform IDE supporting among others the C++ language.

As a template, you could copy the following `qmake.pro` file in your controller directory and open this project in Qt Creator:

> ```qmake

Reviewer: Same comment here, a wizard may be more relevant to generate this `qmake` file.

%tab-component "language"

%tab "C"

```
libraries:
- $(WEBOTS_HOME)/lib/$(SL_PREFIX)Controller$(SL_SUFFIX)

includes:
- $(WEBOTS_HOME)/include/controller/c
```

%tab-end

%tab "C++"

```
libraries:
- $(WEBOTS_HOME)/lib/$(SL_PREFIX)Controller$(SL_SUFFIX)
- $(WEBOTS_HOME)/lib/$(SL_PREFIX)CppController$(SL_SUFFIX)

includes:
- $(WEBOTS_HOME)/include/controller/cpp
```

%tab-end

%tab "Java"

```
libraries:
- $(WEBOTS_HOME)/lib/java/Controller.jar
```

%tab-end

%tab "Python"

```
N/A because Python is interpreted.
```

%tab-end

%end

---

*Legend*

- `WEBOTS_HOME` is the path to Webots.
- `WEBOTS_PROJECT` is the path to your Webots project.
- `CONTROLLER_NAME` is the name of your controller.
- `EXE_EXTENSION` is the executable file suffix.
    - For C or C++: `.exe` on Windows, and an empty string on other operating systems.
    - For Python: `.py`.
    - For Java: `.class` or `.jar`.
- `SL_PREFIX` is the prefix of a shared library: `lib` on Linux or macOS, and an empty string on Windows.
- `SL_SUFFIX` is the suffix of a shared library: `.so` on Linux, `.dylib` on macOS and `.dll` on Windows.
- `PYTHON_VERSION` is your Python version, but concatenated (`27`, `37`, etc.).

---

> Some IDE comes with interpreters or compiler tool chain.

Reviewer: ```suggestion
Some IDE comes with interpreters or a compiler tool chain.
```

Proceed like this: In Webots, hit the `Pause` button then the `Reload` button.
Then, in Visual Studio, use the `Debug / Attach to Process...` menu.
In the dialog choose the `MyController.exe_webots` process.
Still in Visual Studio, you can now add breakpoints and watches in the controller code.
Then, in Webots, hit the `Run` button to resume the simulation.
Now the controller should pause when it reaches one of your breakpoints.

#### Link with the Other Webots Libraries

Webots contains several `C` or `C++` libraries based on the `libController` or `libCppController` libraries (e.g. the `vehicle libraries`, `DARwIn-OP library`, `youBot library`, etc.).

The precompiled `C` libraries are released with their corresponding Visual Studio `.lib` file (the linker to the `.dll` file) to facilitate their integration into a Visual Studio project.
They are located in the same directory as the library source, where the `.dll` file is generated.

However there is no precompiled `C++` libraries for Visual Studio, because the `gcc` compiler tool chain embedded in Webots is incompatible with Visual Studio.

> To use the `C++` libraries with your project, their source files should be compiled directly in your project, exactly as for the `libCppController` library (cf. instructions above).

Reviewer: There are no instructions above about that...

The C++ API is composed of ".hpp" files that contain C++ classes and methods that can be used in C++ controllers only.
In principle any C or C++ controller from the Webots distribution can be turned into a Visual Studio project.

#### Robot Controller Wizard

Since Webots R2018b, a Visual Studio option is offered in the **Wizards / New Robot Controller...** menu after you choose the C or C++ language on Windows.
This wizard creates a Visual Studio project for your robot controller, so that you don't need to configure it manually as described in the next section.

#### Configuration

When creating a Webots controller with Visual Studio, it is necessary to specify the path to Webots ".h" and/or ".hpp" files.
It is also necessary to configure the linker to use the "Controller.lib" import library from Webots distribution.
The "Controller.lib" files is needed to link with the "Controller.dll" file that must be used by the controller in order to communicate with Webots.

The following procedure (Visual Studio 2008 Express) explains how to create a Visual Studio project for a Webots controller.
In this guide, the resulting ".exe" file is launched by Webots.

> Alternatively, Webots can also launch external controllers as explained [here](running-extern-robot-controllers.md#single-simulation-and-single-extern-robot-controller).

Reviewer: ```suggestion
Alternatively, Visual Studio can also launch external controllers as explained [here](running-extern-robot-controllers.md#single-simulation-and-single-extern-robot-controller).
```

Author: Oops.. 😄

The C++ API is composed of ".hpp" files that contain C++ classes and methods that can be used in C++ controllers only.
In principle any C or C++ controller from the Webots distribution can be turned into a Visual Studio project.

#### Robot Controller Wizard

Since Webots R2018b, a Visual Studio option is offered in the **Wizards / New Robot Controller...** menu after you choose the C or C++ language on Windows.
This wizard creates a Visual Studio project for your robot controller, so that you don't need to configure it manually as described in the next section.

#### Configuration

When creating a Webots controller with Visual Studio, it is necessary to specify the path to Webots ".h" and/or ".hpp" files.
It is also necessary to configure the linker to use the "Controller.lib" import library from Webots distribution.
The "Controller.lib" files is needed to link with the "Controller.dll" file that must be used by the controller in order to communicate with Webots.

The following procedure (Visual Studio 2008 Express) explains how to create a Visual Studio project for a Webots controller.

> Note that the resulting ".exe" file must be launched by Webots; it cannot be run from Visual Studio.

Author: I just improved this description here: https://github.com/cyberbotics/webots/pull/372/commits/396b80ad9345b9bced0f4c96742f192373121762
